_posts/2020-07-29-#17977.md

```diff
@@ -0,0 +1,107 @@
+topics this week.**
+
+## Notes
+
+- The _transaction digest_ that is used in signature creation and verification
+  is calculated from different parts of the transaction. There are extensive
+  notes covering this topic from [our session on
+PR18401](https://bitcoincore.reviews/18401.html#pr-18401-initialize-precomputedtransactiondata-in-checkinputscripts).
+
+- BIP 341 contains [the specification of the new signature validation
+  rules](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#signature-validation-rules).
+
+- The [Taproot BIP Review series](https://github.com/ajtowns/taproot-review/)
+  covered this topic in [week
+  4](https://github.com/ajtowns/taproot-review/blob/master/week-4.md#signature-details).
+  [The meeting
+  log](http://www.erisian.com.au/meetbot/taproot-bip-review/2019/taproot-bip-review.2019-11-26-19.01.log.html)
+  from that review session has more details.
+
+- This week, we'll review the implementation of the Taproot transaction hashing
+  algorithm. We won't look at full signature validation. It may be useful to
+  checkout just up to the commit _Implement Taproot signature hashing (BIP 341)_.
+  Specifically, we'll be looking at the `SignatureHashSchnorr()` function in
+  `src/script/interpreter.cpp`.
+
+- It's probably most instructive to compare the new `SignatureHashSchnorr()`
+  function to the `SignatureHash()` function immediately below, which
+  calucalates the signature hash for segwit v0 and legacy transactions.
+
+## Questions
+
+1. `SignatureHashSchnorr()` is a templated function. Why?  What types `T` is
+   the template function instantiated with? _Hint: look at how the existing
+   `SignatureHash()` function is called._
+
+2. `SignatureHashSchnorr()` is passed a `PrecomputedTransactionData*` argument.
+   What data is stored in this structure? Why?
+
+3. `SignatureHashSchnorr()` is passed a `hash_type` argument. How many valid
+   hash types are there? The `hash_type` parameter is split into `output_type`
+   and `intput_type` here:
+
+     ```
+     // Hash type
+     const uint8_t output_type = ...
+     const uint8_t input_type = ...
+     ```
+
+     How many valid `output_type`s and `intput_type`s are there?
+
+4. Just like `SignatureHash()`, the new function creates a local `CHashWriter`
+   object. `CHashWriter` is a stream-like object. Other objects can be
+   serialized into it (using the `<<` operator), and at the end,
+   `CHashWriter::GetHash()` is called to retrieve the digest.
+
+     One difference from `SignatureHash()` is that the `CHashWriter` is
+     copy-constructed from `HasherTapSighash`. How is _that_ object constructed, and
+     what's the difference from a regular `CHashWriter` object?
+
+5. If the `hash_type` does not have the `ANYONECANPAY` flag, certain parts of
+   the transaction are added to the `CHashWriter`. What are those elements, and
+   how is that different from in `SignatureHash()`?
```

Comment:
"from in" -> maybe just one of the two words

Reply:
done